akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -0,0 +1,218 @@
+  }
+
+  def senderActorRef: String = {
+    if (_senderActorRef ne null) _senderActorRef
+    else {
+      _senderActorRef = compressionTable.decompressActorRef(_senderActorRefIdx)
+      _senderActorRef
+    }
+  }
+
+  def recipientActorRef_=(ref: String): Unit = {
+    _recipientActorRef = ref
+    _recipientActorRefIdx = compressionTable.compressActorRef(ref)
+  }
+
+  def recipientActorRef: String = {
+    if (_recipientActorRef ne null) _recipientActorRef
+    else {
+      _recipientActorRef = compressionTable.decompressActorRef(_recipientActorRefIdx)
+      _recipientActorRef
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class EnvelopeBuffer(val byteBuffer: ByteBuffer) {
+  import EnvelopeBuffer._
+
+  def writeHeader(h: HeaderBuilder): Unit = {
+    val header = h.asInstanceOf[HeaderBuilderImpl]
+    byteBuffer.clear()
+
+    // Write fixed length parts
+    byteBuffer.putInt(header.version)
+    byteBuffer.putInt(header.uid)
+
+    // Write compressable, variable-length parts always to the actual position of the buffer
+    // Write tag values explicitly in their proper offset
+    byteBuffer.position(LiteralsSectionOffset)
+
+    // Serialize sender
+    if (header._senderActorRefIdx != -1)
+      byteBuffer.putInt(SenderActorRefTagOffset, header._senderActorRefIdx | TagTypeMask)
+    else
+      writeLiteral(SenderActorRefTagOffset, header.senderActorRef)
+
+    // Serialize recipient
+    if (header._recipientActorRefIdx != -1)
+      byteBuffer.putInt(RecipientActorRefTagOffset, header._recipientActorRefIdx | TagTypeMask)
+    else
+      writeLiteral(RecipientActorRefTagOffset, header.recipientActorRef)
+  }
+
+  def parseHeader(h: HeaderBuilder): Unit = {
+    val header = h.asInstanceOf[HeaderBuilderImpl]
+    byteBuffer.flip()
+
+    // Read fixed length parts
+    header.version = byteBuffer.getInt
+    header.uid = byteBuffer.getInt
+
+    // Read compressable, variable-length parts always from the actual position of the buffer
+    // Read tag values explicitly from their proper offset
+    byteBuffer.position(LiteralsSectionOffset)
+
+    // Deserialize sender
+    val senderTag = byteBuffer.getInt(SenderActorRefTagOffset)
+    if ((senderTag & TagTypeMask) != 0) {
+      val idx = senderTag & TagValueMask
+      header._senderActorRef = null
+      header._senderActorRefIdx = idx
+    } else {
+      header._senderActorRef = readLiteral()
+    }
+
+    // Deserialize recipient
+    val recipientTag = byteBuffer.getInt(RecipientActorRefTagOffset)
+    if ((recipientTag & TagTypeMask) != 0) {
+      val idx = recipientTag & TagValueMask
+      header._recipientActorRef = null
+      header._recipientActorRefIdx = idx
+    } else {
+      header._recipientActorRef = readLiteral()
+    }
+  }
+
+  private def readLiteral(): String = {
+    val length = byteBuffer.getShort
+    val bytes = Array.ofDim[Byte](length)
+    byteBuffer.get(bytes)
+    new String(bytes, UsAscii)
```

Comment:
Heh, so waiting for the day when String can be constructed from a ByteBuffer range...